### ISSUE TITLE
Assignment of `if` result incorrectly autocorrected

### ISSUE THREAD

**Issue by amw** [bbatsov/rubocop#2323]
I got this bad autocorrect. The original code:

``` ruby
  def expression= value
    self[:expression_id] =
      if value.kind_of? Expression
        value.id
      else
        nil
      end
  end
```

Was changed to:

``` ruby
  def expression= value
    self[:expression_id] =
      value.id if value.is_a? Expression
end
```

This of course is incompatible. It makes the whole assignment conditional.

If you think this has to be corrected at all then you should use ternary operator:

``` ruby
  def expression= value
    self[:expression_id] = value.is_a?(Expression) ? value.id : nil
  end
```

**Comment by alexdowad** [comment]
This is caused by the `Style/IfUnlessModifier` cop. `Style/EmptyElse` is also involved, but the transform it does is valid.

**Comment by swatosh** [comment]
Unless self[:expression_id] has a value that should be overwritten with nil

**Comment by alexdowad** [comment]
Sorry, @swatosh, what do you mean?

**Comment by swatosh** [comment]
If self[:expression_id == 1234 before the method call, there is a code path
in the original that would change it to nil.

Sorry, @swatosh <https://github.com/swatosh>, what do you mean?

—
Reply to this email directly or view it on GitHub
<https://github.com/bbatsov/rubocop/issues/2323#issuecomment-151522464>.

**Comment by alexdowad** [comment]
> If self[:expression_id == 1234 before the method call, there is a code path
> in the original that would change it to nil.

Yes. That is the whole point of the linked PR. 

Currently, RuboCop removes the code path which sets `self[:expression_id]` to `nil`. With the linked PR merged, it will not do so.

**Comment by swatosh** [comment]
Sorry, I missed the PR. I was responding to "...but the transform it does
is valid." Again, sorry for misunderstanding.

**Comment by amw** [comment]
@swatosh I think it meant that `Style/EmptyElse` transform is valid and only `Style/IfUnlessModifier` needs fixing.
